### ISSUE TITLE
context: `RemoteAddr()` is unsafe

### ISSUE THREAD

**Issue by wuhan005** [flamego/flamego#115]
### Component

Core framework

### Middleware

n/a

### Version

1.1.0

### Describe the bug

As the document mentioned, the `RemoteAddr()` method looks at the following things in the order to determine which one is more likely to contain the real client address.
> The X-Real-IP request header
The X-Forwarded-For request header
The http.Request.RemoteAddr field

So the client's IP can be spoofed by setting the `X-Real-IP` or `X-Forwarded-For` header when the Flamego server is exposed directly to the internet.

The Gin web framework has the same vulnerability before, check CVE-2020-28483 for details.

I think we should create a new function called `ClientIP()` to get the client IP address from these headers. The `RemoteAddr()` can only get it from `http.Request.RemoteAddr`.

Moreover, the `Flame` instance should allow the developers to set the trusted remote IP headers and the trusted proxies if they want to get the client IP from user-provided headers.

### To reproduce

n/a

### Expected behavior

n/a

### Additional context

https://nvd.nist.gov/vuln/detail/CVE-2020-28483
https://github.com/gin-gonic/gin/issues/2473

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by unknwon** [comment]
This is known and warning has been put in the docs:

<img width="855" alt="CleanShot 2022-04-04 at 14 58 50@2x" src="https://user-images.githubusercontent.com/2946214/161489972-712ce582-486e-4288-9ca1-e55c88980ea9.png">

> I think we should create a new function called `ClientIP()` to get the client IP address from these headers. The `RemoteAddr()` can only get it from `http.Request.RemoteAddr`.

If the user just want the `http.Request.RemoteAddr`, then they can just call this exact method/field.

Meanwhile, I think if the user wants custom logic, they can just wrap the `flamego.Context` and implement/override their own version of this method.



> Moreover, the `Flame` instance should allow the developers to set the trusted remote IP headers and the trusted proxies if they want to get the client IP from user-provided headers.

We can provide examples, but I don't think it is worth to complicate the core framework. WDYT?

**Comment by wuhan005** [comment]
I agree with you, users can implement the `ClientIP()` method in their custom context.

Maybe we can make a more clear warning message to point out there is *a potential security issue*. The current warning only shows proxy or VPN can fake the IP address, but the attacker can simply change the request header to do so as well.

**Comment by unknwon** [comment]
> Maybe we can make a more clear warning message to point out there is _a potential security issue_. The current warning only shows proxy or VPN can fake the IP address, but the attacker can simply change the request header to do so as well.

Sounds good, feel free to send PRs to docs repos.
